denote.el

```diff
@@ -3933,6 +3937,143 @@ To be assigned to `markdown-follow-link-functions'."
 (eval-after-load 'markdown-mode
   '(add-hook 'markdown-follow-link-functions #'denote-link-markdown-follow))
 
+;;;;; Link fontification
+(defcustom denote-link-display-style 'button
+  "The style for displaying denote links.
+
+By default (`button'), the link is displayed as a button. This
+requires calling `denote-link-buttonize-buffer' or enabling the
+minor-mode `denote-fontify-links-mode' appropriately. See the
+documentation of that function for more details.
+
+When the value is `fontify', the link is fontified using
+font-lock. The appropriate keyword must be added to `font-lock'.
+Enabling the minor-mode `denote-fontify-links-mode' does this.
+
+The value `fontify-and-fold', is the same as `fontify' but also
+uses `org-fold' (if available) to fold the links.
+"
+  :type '(choice
+          (const :tag "As button" button)
+          (const :tag "Fontified" fontify)
+          (const :tag "Fontified and folded" fontify-and-fold))
+  :package-version '(denote . "0.6.0")
+  :group 'denote)
+
+(defvar denote-link-keymap
+  (let ((map (make-sparse-keymap)))
+    (define-key map [mouse-2] #'denote-link-open-at-mouse)
+    (define-key map [mouse-3] #'denote-link-open-at-mouse)
+    (define-key map [follow-link] #'mouse-face)
+    map)
+  "Keymap used in fontified denote links")
+
+(defun denote-link-open-at-mouse (ev)
+  "Open denote link after mouse click."
+  (interactive "e")
+  (mouse-set-point ev)
+  (when-let ((id (get-text-property (point) 'denote-link-id)))
+    (funcall denote-link-button-action (denote-get-path-by-id id))))
+
+(defun denote-fontify-links (&optional limit)
+  "Fontify denote links.
+
+Implementation based on `org-activate-links--text-properties'"
+  (catch :exit
+    (when-let (type (denote-filetype-heuristics (buffer-file-name)))
+      (while (re-search-forward
+              (denote--link-in-context-regexp type)
+              limit t)
+        (save-match-data  ;; to return the matches to font-lock
+          (let* ((start (match-beginning 0))
+	         (end (match-end 0))
+	         (visible-start (match-beginning 2))
+	         (visible-end (match-end 2))
+                 (id (match-string-no-properties 1))
+                 (link (concat "file:" (denote-get-path-by-id id))))
+	    (unless (let ((face (get-text-property
+			         (max (1- start) (point-min)) 'face)))
+		      (if (consp face)
+                          (memq 'font-lock-comment-face face)
+		        (eq 'font-lock-comment-face face)))
+	      (when (and (featurep 'org-fold)
+                         (eq denote-link-display-style 'fontify-and-fold))
+                (org-fold-region start end nil 'org-link)
```

Comment:
The byte compiler is warning me that `org-fold` requires Emacs 29 to work. Is this correct? We now depend on Emacs 28 and I hope we keep it this way for the foreseeable future (let's say until the next Debian stable, which is about a year away).

Reply:
I am not sure, but note that I only enable the folding features when `org-fold` is available as a feature.